pennylane/qnode.py

```diff
@@ -558,45 +558,58 @@ def _best_method(self, idx):
         #
         # 5. Then run the standard discrete-case algorithm for determining the best gradient method
         # for every free parameter.
-        def best_for_op(o_idx):
+
+        # pylint: disable=too-many-return-statements
+        def best_for_op(op_node):
             "Returns the best gradient method for the operation op."
-            op = self.ops[o_idx]
             # for discrete operations, other ops do not affect the choice
+            o_idx = op_node["idx"]
+            op = op_node["op"]
+
             if not isinstance(op, qml.operation.CV):
                 return op.grad_method
 
             # for CV ops it is more complicated
-            if op.grad_method == 'A':
+            if op.grad_method == "A":
                 # op is Gaussian and has the heisenberg_* methods
-                # check that all successor ops are also Gaussian
-                # TODO when we upgrade to a DAG: a non-Gaussian successor is OK if it
-                # isn't succeeded by any observables?
-                successors = self._op_successors(o_idx, 'G')
-
-                if all(x.supports_heisenberg for x in successors):
-                    # check successor EVs, if any order-2 observables are found return 'A2', else return 'A'
-                    ev_successors = self._op_successors(o_idx, 'E')
-                    for x in ev_successors:
-                        if x.ev_order is None:
-                            return 'F'
-                        if x.ev_order == 2:
-                            if x.return_type is qml.operation.Variance:
-                                # second order observables don't support
-                                # analytic diff of variances
-                                return 'F'
-                            op.grad_method = 'A2'  # bit of a hack
+
+                obs_successors = self._op_successors(o_idx, 'E')
+                if not obs_successors:
+                    # op is not succeeded by any observables, thus analytic method is OK
                     return 'A'
 
-                return 'F'
+                # check that all successor ops are also Gaussian
+                successor_ops = self._op_successors(o_idx, 'G', get_nodes=True)
+                if not all(x["op"].supports_heisenberg for x in successor_ops):
+                    non_gaussian_ops = [x for x in successor_ops if not x["op"].supports_heisenberg]
```

Comment:
Just a thought for later, but we probably want to change "supports_heisenberg" to just be "is_gaussian"